### PR TITLE
Add a SerializationContextEvent to the API.

### PR DESCRIPTION
**Pull request by mallezie**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [X]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [NO ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

#### Description:

This PR adds an Event to the API on the moment (before and after) the serialization context is set. This allows event listeners to change the context, and change the existing exclusion groups and / or strategies to alter the output from the API (and limit it for example). 

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Implement a listener which can alter the serialization context.

### CODE REVIEW

File: app/bundles/ApiBundle/Event/ApiSerializationContextEvent.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+namespace Mautic\ApiBundle\Event;
+
+use FOS\RestBundle\Context\Context;
+use Mautic\CoreBundle\Event\CommonEvent;
+use Symfony\Component\HttpFoundation\Request;
+
+class ApiSerializationContextEvent extends CommonEvent
+{
+    /**
+     * @varContext
```
Comment (mollux) [inline app/bundles/ApiBundle/Event/ApiSerializationContextEvent.php]: @mallezie this looks weird :)

Comment (mallezie) [reply app/bundles/ApiBundle/Event/ApiSerializationContextEvent.php]: Removed the comment, since it's a fully typed property.

**Comment by RCheesley** [comment]
@all-contributors please add @mallezie  for code

Thanks for making the PR @mallezie  -  welcome to the community!

File: app/bundles/ApiBundle/ApiEvents.php [hunk 2]
```diff
@@ -60,4 +60,22 @@ final class ApiEvents
      * @var string
      */
     public const API_ON_ENTITY_POST_SAVE = 'mautic.api_on_entity_post_save';
+
+    /**
+     * The mautic.api_pre_serialization_context event is thrown before the serialization context is created for the view.
```
Comment (mollux) [inline app/bundles/ApiBundle/ApiEvents.php]: the word `thrown` sound weird here (i know the other events in this and other files have similar comments)
```suggestion
     * The mautic.api_pre_serialization_context event is dispatched before the serialization context is created for the view.
```

File: app/bundles/ApiBundle/ApiEvents.php [hunk 3]
```diff
@@ -60,4 +60,22 @@ final class ApiEvents
...
+     *
+     * The event listener receives a Mautic\ApiBundle\Event\ApiSerializationContextEvent instance.
+     *
+     * @var string
+     */
+    public const API_PRE_SERIALIZATION_CONTEXT = 'mautic.api_pre_serialization_context';
+
+    /**
+     * The mautic.api_post_serialization_context event is thrown after the serialization context is created for the view.
```
Comment (mollux) [inline app/bundles/ApiBundle/ApiEvents.php]: the word `thrown` sound weird here (i know the other events in this and other files have similar comments)
```suggestion
     * The mautic.api_post_serialization_context event is dispatched after the serialization context is created for the view.
```

**Review by mollux — COMMENTED** [review]
@mallezie looks good, some minor nitpicks about the comments

**Comment by mallezie** [comment]
Updated the comments.

**Review by mollux — DISMISSED** [review]
looks good to me!

**Comment by RCheesley** [comment]
Could we please resolve the conversations if they have been addressed in the commits @mallezie / @mollux?  Many thanks!

**Comment by mallezie** [comment]
@RCheesley I've resolved the fixed conversations.

**Comment by RCheesley** [comment]
Wonderful, thanks for getting that done so quickly!

**Comment by RCheesley** [comment]
It looks like we might need to bump up the code coverage on this PR. Is that something you might be willing to take a look at @mallezie ?

**Comment by RCheesley** [comment]
@mallezie is there any way you can give more detail in the description to help testers? I know it's a fairly complex one to test, but maybe you could share a sample plugin or some code that could be used, and the API commands to be issued? It's more likely to get tested that way.

**Comment by mallezie** [comment]
Update the branch with upstream. 
For adding test coverage, this adds a new event, which has no logic on itself, so no sure how to add test coverage for it. I see some events have tests, but those are events having logic. Mostly event listeners are tested, but there is no eventlistener to this event in core. 
For manual testing, i've added a plugin which subscribes to this event. It changes the API result to only show the id-field and not the rest of the data. 

[SerializationTestBundle.zip](https://github.com/mautic/mautic/files/14110436/SerializationTestBundle.zip)

You can test this now, by installing the attached plugin, and see the results of an API changed. 
* Install M5
* Create a contact
* Install plugin
* Do a API query. 
```
curl -s -u"admin":"Maut1cR0cks!" https://mautic5.ddev.site/api/contacts/1 | jq
{
  "contact": {
    "id": 1
  }
}
```

**Comment by escopecz** [comment]
The Rector check is failing. Could you please take a look?

File: app/bundles/ApiBundle/Event/ApiSerializationContextEvent.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
+<?php
+
```
Comment (escopecz) [inline app/bundles/ApiBundle/Event/ApiSerializationContextEvent.php:2]:
```suggestion

declare(strict_types=1);

```

File: app/bundles/ApiBundle/Event/ApiSerializationContextEvent.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+namespace Mautic\ApiBundle\Event;
+
+use FOS\RestBundle\Context\Context;
+use Mautic\CoreBundle\Event\CommonEvent;
+use Symfony\Component\HttpFoundation\Request;
+
+class ApiSerializationContextEvent extends CommonEvent
```
Comment (escopecz) [inline app/bundles/ApiBundle/Event/ApiSerializationContextEvent.php]:
```suggestion
final class ApiSerializationContextEvent extends CommonEvent
```
There is no need for a DTO object to be extensible.

File: app/bundles/ApiBundle/Event/ApiSerializationContextEvent.php [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+{
+    protected Context $context;
+
+    private Request $request;
+
+    public function __construct(Context $context, Request $request)
+    {
+        $this->context                 = $context;
+        $this->request                 = $request;
+    }
```
Comment (escopecz) [inline app/bundles/ApiBundle/Event/ApiSerializationContextEvent.php]:
```suggestion
    public function __construct(private Context $context, private Request $request)
    {
    }
```

File: app/bundles/ApiBundle/Event/ApiSerializationContextEvent.php [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+    public function getContext(): Context
+    {
+        return $this->context;
+    }
+
+    /**
+     * Sets the context.
+     */
```
Comment (escopecz) [inline app/bundles/ApiBundle/Event/ApiSerializationContextEvent.php]:
```suggestion
```
The comment just states the obvious. It's unnecessary.

**Review by escopecz — CHANGES_REQUESTED** [review]

**Comment by mallezie** [comment]
Processed all feedback, and rector is now passing.

**Review by escopecz — APPROVED** [review]
Thank you! 👍

**Pull request merged**